operators/pkg/controller/elasticsearch/driver/downscale.go

```diff
@@ -14,122 +15,221 @@ import (
+	// clients
+	k8sClient k8s.Client
+	esClient  esclient.Client
+	// driver states
+	resourcesState reconcile.ResourcesState
+	observedState  observer.State
+	reconcileState *reconcile.State
+	expectations   *reconciler.Expectations
+	// ES cluster
+	es v1alpha1.Elasticsearch
+}
+
+// HandleDownscale attempts to downscale actual StatefulSets towards expected ones.
+func HandleDownscale(
+	downscaleCtx downscaleContext,
 	expectedStatefulSets sset.StatefulSetList,
 	actualStatefulSets sset.StatefulSetList,
-	esClient esclient.Client,
-	resourcesState reconcile.ResourcesState,
-	observedState observer.State,
-	reconcileState *reconcile.State,
 ) *reconciler.Results {
 	results := &reconciler.Results{}
 
-	// compute the list of nodes leaving the cluster, from which
-	// data should be migrated away
-	leavingNodes := []string{}
+	// compute the list of StatefulSet downscales to perform
+	downscales := calculateDownscales(expectedStatefulSets, actualStatefulSets)
+	leavingNodes := leavingNodeNames(downscales)
 
-	// process each statefulset for downscale
-	for i, actual := range actualStatefulSets {
-		expected, shouldExist := expectedStatefulSets.GetByName(actual.Name)
-		targetReplicas := int32(0) // sset removal
-		if shouldExist {           // sset downscale
-			targetReplicas = sset.Replicas(expected)
+	// migrate data away from nodes that should be removed
+	if err := scheduleDataMigrations(downscaleCtx.esClient, leavingNodes); err != nil {
+		return results.WithError(err)
+	}
+
+	for _, downscale := range downscales {
+		// attempt the StatefulSet downscale (may or may not remove nodes)
+		requeue, err := attemptDownscale(downscaleCtx, downscale, leavingNodes, actualStatefulSets)
+		if err != nil {
+			return results.WithError(err)
 		}
-		leaving, removalResult := d.scaleStatefulSetDown(actualStatefulSets, &actualStatefulSets[i], targetReplicas, esClient, resourcesState, observedState, reconcileState)
-		results.WithResults(removalResult)
-		if removalResult.HasError() {
-			return results
+		if requeue {
+			// retry downscaling this statefulset later
+			results.WithResult(defaultRequeue)
 		}
-		leavingNodes = append(leavingNodes, leaving...)
 	}
 
-	// migrate data away from nodes leaving the cluster
-	log.V(1).Info("Migrating data away from nodes", "nodes", leavingNodes)
-	if err := migration.MigrateData(esClient, leavingNodes); err != nil {
-		return results.WithError(err)
+	return results
+}
+
+// ssetDownscale helps with the downscale of a single StatefulSet.
+type ssetDownscale struct {
+	statefulSet     appsv1.StatefulSet
+	initialReplicas int32
+	targetReplicas  int32
+}
+
+// leavingNodeNames returns names of the nodes that are supposed to leave the Elasticsearch cluster
+// for this StatefulSet.
+func (d ssetDownscale) leavingNodeNames() []string {
+	if d.targetReplicas >= d.initialReplicas {
+		return nil
+	}
+	leavingNodes := make([]string, 0, d.initialReplicas-d.targetReplicas)
+	// nodes are ordered by highest ordinal first
```

Comment:
nit: I'd move that to godoc for this func, seems important for the rest of the code.